### PR TITLE
Added password support for all archivers

### PR DESCRIPTION
**Pull request by Swich1987**
Added password support for all archivers, which support work with password: arc, arj, p7zip, py_zipfile, rar, unace, unalz, unzip, xdms. Added tests for those, which support creation and test files in data folder for it.

I checked every manual for all programs on my Ubuntu and on the Internet, if I find support for password, I added it. I added new tests for every changed programs, and added test archive for them if it's possible to create. For example I can't add test files for unace, unalz, xdms and etc. I can make it, if somebody send me archived files encrypted with password "thereisnotry" (I choose it as default for tests).

It's my first open source commit, I'm pretty new in it, so will do my best to answer any comments ASAP.

### CODE REVIEW

File: patoolib/__init__.py
Comment (yarikoptic) [inline patoolib/__init__.py:639]: this looks suboptimal (sure it would never be `"password"` instead of `'password'` and message itself does not change?) -- if we are to rely on keyword arg to be known, may be some inspection to be done instead of catching/analyzing TypeError?  and don't we already have that information in NoPasswordSupportArchivePrograms?

File: patoolib/programs/arc.py
Comment (yarikoptic) [inline patoolib/programs/arc.py:25]: I with there was `PatoolValueError`, molding all exceptions as `PatoolError` (there is no other) makes outside handling trickier.  But I guess it should not be this PR to introduce it

**Review by yarikoptic — COMMENTED** [review]
Thanks!  it is touching many files, but otherwise looks straightforward and worthwhile addition I think.  Left one question/comment

File: patoolib/__init__.py
Comment (Swich1987) [reply patoolib/__init__.py:639]: Yes, I thought about it, and not sure about better solution. I tried to realize it in pythonic way: try something, than catch error. NoPasswordSupportArchivePrograms was made as double of ArchivePrograms, to get subset of it, which support password in archives. This way you can just add new archive programs to AchivePrograms, set password argument and it will work.

I can make, instead, PasswordSupportArchivePrograms, which will realize exact that subset. Minus here is that anyone, who would add some program with password support, will need to add it to both dicts: ArchivePrograms and PasswordSupportArchivePrograms. Will it be better than current realization?

File: patoolib/programs/arc.py
Comment (Swich1987) [reply patoolib/programs/arc.py:25]: It's not a big deal to create new exception, much harder is to make correct decision about where to use it in logical and straightforward way. 
I can try to implement it, or later create new PR, from my point of view it's much simpler than adding password 😄

File: patoolib/__init__.py
Comment (yarikoptic) [reply patoolib/__init__.py:639]: what about `'password' in inspect.signature(archive_cmdlist_func).parameters` ?

Comment (Swich1987) [reply patoolib/__init__.py:639]: Yes, I think it will be much better. I will change it now.

Comment (Swich1987) [reply patoolib/__init__.py]: Changed

File: patoolib/programs/arc.py
Comment (Swich1987) [reply patoolib/programs/arc.py:25]: I will add it to new PR after merge this.

File: patoolib/__init__.py
Comment (yarikoptic) [inline patoolib/__init__.py:633]: minor: could be just 

```suggestion
            if kwargs.get('password') is None:
```
to avoid duplication, and make code shorter/even more readable

File: patoolib/programs/p7zip.py
Comment (yarikoptic) [inline patoolib/programs/p7zip.py:118]: these two lines are common to many invocations. Having a helper

```
def _maybe_add_password(cmdline, password):
    if password:
        cmdlist.append('-p%s' % password)
```
and use it instead would allow to centralize the logic and thus in the future possibly add a single location to fix up/extend (if we discover that some characters aren't legit etc)

Comment (yarikoptic) [reply patoolib/programs/p7zip.py:118]: similarly for other tools as well

**Review by yarikoptic — APPROVED** [review]
looks good to me. left some comments how to make better. even if not addressed, ok with me for now, and will merge in a day or two if no objections are voiced

File: patoolib/__init__.py
Comment (Swich1987) [reply patoolib/__init__.py:633]: It will be True when there is no 'password' in kwargs, so pop will raise exception.
Of course, I can use suppress, but it will be one more line, so I'm not sure it'll be better.

File: patoolib/programs/p7zip.py
Comment (Swich1987) [reply patoolib/programs/p7zip.py:118]: Agreed, it's better. As I understood you propose separate helper for every program. Maybe will be better to add some global helper for all programs? Something like this:
```
def _maybe_add_password(cmdlist, password, added_args):
    if password:
        if isinstance(added_args, list):
            cmdlist.extend(added_args)
        else:
            cmdlist.append(added_args)
```
and use it like:
```
from . import maybe_add_password
maybe_add_password(cmdlist, password, '-p%s' % password)
```
or 
```
maybe_add_password(cmdlist, password, ['-p', password])
```

File: patoolib/__init__.py
Comment (yarikoptic) [reply patoolib/__init__.py:633]: ah, right, I just misread the logic - my bad.  I think then it is ok as is.

FWIW, it could be `.pop("password", None)` to avoid an exception if not there.

File: patoolib/programs/p7zip.py
Comment (yarikoptic) [reply patoolib/programs/p7zip.py:118]: yeah, I was also considering generic helper, but decided against it -- it would indeed then duplicate all the specifications of password args for each invocation, thus not much really of help.  So probably sticking with a dedicated for each program would be of the most benefit

Comment (Swich1987) [reply patoolib/programs/p7zip.py]: Changed

**Comment by yarikoptic** [comment]
we need to remove python2 first for that `.signature`. Submitted #98 , then when/if merged CI should be rerun for this PR as well

**Comment by yarikoptic** [comment]
please push some rewrite or changes to re-trigger CI

**Comment by Swich1987** [comment]
amended last commit and pushed, hope that re-trigger CI

**Comment by Swich1987** [comment]
Yeah, I see it, thanks to your link in #98. So it's green now:
https://travis-ci.org/github/wummel/patool/builds/759448115

**Pull request merged**
